### PR TITLE
Add attribute support to Home Assistant sensors

### PR DESCRIPTION
**Pull request by frenck**
# What does this implement/fix? 

This PR adds support for specifying the Home Assistant entity attribute on ESPHome `homeassistant` sensors. This allows for tracking Home Assistant entity state attributes in ESPHome (besides the existing entity state).

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Configuration change (this will require users to update their yaml configuration files to keep working)

**Related issue or feature (if applicable):** esphome/feature-requests#1112

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#1159
  
# Test Environment

- [x] ESP32
- [x] ESP8266
- [ ] Windows
- [x] Mac OS
- [x] Linux

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
sensor:
  - platform: homeassistant
    id: the_temperature
    entity_id: climate.living_room
    attribute: current_temperature
    
binary_sensor:
  - platform: homeassistant
    id: occupancy
    entity_id: binary_sensor.kitchen_motion
    attribute: occupancy
    
  # Same light as in sensor, to ensure callbacks aren't mixed up
  - platform: homeassistant
    id: the_temperature
    entity_id: climate.living_room

text_sensor:
  - platform: homeassistant
    id: current_effect
    entity_id: light.led_strip
    attribute: effect
```

# Explain your changes

Describe your changes here to communicate to the maintainers **why we should accept this pull request**.

🤷 See above. It requested a lot.

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

## Additional information

This PR needs changes to Home Assistant as well, however, this depends on the following PR in `aioesphomelib`: <https://github.com/esphome/aioesphomeapi/pull/30>

## TODO

This stuff I created in the middle of the night and I really need to sit down to look at it when I'm awake more. Some stuff I think I still should do (besides re-reading the stuff I dumped here):

- [x] Format generated files... (wut? Why isn't that done automatically?!)
- [x] Test with ESP8266
- [x] Test with older HA instance, verify backward compatibility
- [x] Test with sensors
- [x] Test with binary sensors
- [x] Test with text sensors
- [x] Open up docs PR 
- [x] Open up PR on HA side to bump `aioesphomelib`: <https://github.com/home-assistant/core/pull/50522>
- [x] Open up PR on HA side to add support for state attribute tracking: <https://github.com/home-assistant/core/pull/50528>
- [x] Fix proto spec out of sync (aioesphome<->esphome)
- [x] Fix custom API (backward compatibility)

### CODE REVIEW

File: esphome/components/homeassistant/binary_sensor/homeassistant_binary_sensor.cpp [hunk 1]
```diff
@@ -31,6 +36,11 @@ void HomeassistantBinarySensor::setup() {
 void HomeassistantBinarySensor::dump_config() {
   LOG_BINARY_SENSOR("", "Homeassistant Binary Sensor", this);
   ESP_LOGCONFIG(TAG, "  Entity ID: '%s'", this->entity_id_.c_str());
+  if (this->attribute_.has_value()) {
+    ESP_LOGCONFIG(TAG, "  Attribute: '%s'", this->attribute_.value().c_str());
+  } else {
+    ESP_LOGCONFIG(TAG, "  No Attribute");
```
Comment (glmnet) [inline esphome/components/homeassistant/binary_sensor/homeassistant_binary_sensor.cpp:42]: This might be confusing. I'd log nothing

**Review by glmnet — COMMENTED** [review]
Looks good 🚂

File: esphome/components/api/custom_api_device.h [hunk 2]
```diff
@@ -76,13 +76,13 @@ class CustomAPIDevice {
     global_api_server->register_user_service(service);
   }
 
-  /** Subscribe to the state of an entity from Home Assistant.
+  /** Subscribe to the state (or attritebute state) of an entity  from Home Assistant.
```
Comment (jesserockz) [inline esphome/components/api/custom_api_device.h:79]:
```suggestion
  /** Subscribe to the state (or attribute state) of an entity from Home Assistant.
```

**Review by jesserockz — APPROVED** [review]
Awesome. Thanks

**Comment by frenck** [comment]
Downstream PRs @ Home Assistant have been merged (for 2021.6)

File: esphome/components/api/api_connection.cpp [hunk 3]
```diff
@@ -642,8 +642,9 @@ DeviceInfoResponse APIConnection::device_info(const DeviceInfoRequest &msg) {
 }
 void APIConnection::on_home_assistant_state_response(const HomeAssistantStateResponse &msg) {
   for (auto &it : this->parent_->get_state_subs())
-    if (it.entity_id == msg.entity_id)
+    if (it.entity_id == msg.entity_id and it.attribute.value() == msg.attribute) {
```
Comment (jesserockz) [inline esphome/components/api/api_connection.cpp:645]: I don't believe we use `and` anywhere else in the codebase. Maybe change for consistency?
```suggestion
    if (it.entity_id == msg.entity_id && it.attribute.value() == msg.attribute) {
```

Comment (frenck) [reply esphome/components/api/api_connection.cpp]: I blame too much Python in my case 😬

Comment (jesserockz) [reply esphome/components/api/api_connection.cpp]: Haha, yeah I had to double check the file I was reading

File: tests/custom.h [hunk 4]
```diff
@@ -49,6 +49,7 @@ class CustomNativeAPI : public Component, public CustomAPIDevice {
     register_service(&CustomNativeAPI::on_start_dryer, "start_dryer", {"value"});
     register_service(&CustomNativeAPI::on_many_values, "many_values", {"bool", "int", "float", "str1", "str2"});
     subscribe_homeassistant_state(&CustomNativeAPI::on_light_state, "light.my_light");
+    subscribe_homeassistant_state(&CustomNativeAPI::on_light_state, "light.my_light", "brightness");
```
Comment (jesserockz) [inline tests/custom.h:52]: Copy paste?
```suggestion
    subscribe_homeassistant_state(&CustomNativeAPI::on_brightness_state, "light.my_light", "brightness");
```

**Review by jesserockz — APPROVED** [review]

**Pull request merged**